### PR TITLE
VecZ + MatZ Mul, Cmp, Transpose

### PR DESCRIPTION
**Pull request by jnsiemer**
This PR implements
- [X] Compare
- [X] Tranpose
- [X] Multiplication
including tests for `MatZ`.

Additionally
- [X] new and from_str
- [X] getter and setter
- [X] Multiplication (also with MatZ)
- [X] transponse
including tests are implemented for `VecZ`.

### CODE REVIEW

File: src/integer/mat_z/mul.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    traits::{GetNumColumns, GetNumRows},
+};
+use flint_sys::fmpz_mat::fmpz_mat_mul;
+use std::ops::Mul;
+
+impl Mul for &MatZ {
+    type Output = MatZ;
+
+    /// Implements the [`Mul`] trait for two [`MatZ`] values.
+    /// [`Mul`] is implemented for any combination of owned and borrowed [`MatZ`] and any kind of [`VecZ`].
```
Comment (Marvin-Beckmann) [inline src/integer/mat_z/mul.rs:27]: VecZ? 
```suggestion
    /// [`Mul`] is implemented for any combination of owned and borrowed [`MatZ`].
```

File: src/integer/mat_z/mul.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    ///
+    /// let c = &a * &b;
+    /// let d = a * b;
+    /// let e = &c * d;
+    /// let f = c * &e;
+    /// ```
+    fn mul(self, other: Self) -> Self::Output {
+        if self.get_num_columns() != other.get_num_rows() {
+            panic!("Matrix dimensions do not match for matrix multiplication!");
+        }
```
Comment (Marvin-Beckmann) [inline src/integer/mat_z/mul.rs:50]:
```suggestion
        // TODO: Call mul_safe
        if self.get_num_columns() != other.get_num_rows() {
            panic!("Matrix dimensions do not match for matrix multiplication!");
        }
```

**Review by Marvin-Beckmann — CHANGES_REQUESTED** [review]

File: src/integer/mat_z/mul.rs (hunk 1 above)
Comment (jnsiemer) [reply src/integer/mat_z/mul.rs:27]: This comment is correct. It's about all the combinations of Multiplication of that type that are implemented.

File: src/integer/mat_z/mul.rs (hunk 2 above)
Comment (jnsiemer) [reply src/integer/mat_z/mul.rs:51]: Done

File: src/integer/mat_z/mul.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+        let mut new = MatZ::new(self.get_num_rows(), other.get_num_columns()).unwrap();
+        unsafe { fmpz_mat_mul(&mut new.matrix, &self.matrix, &other.matrix) };
+        new
+    }
+}
+
+arithmetic_trait_borrowed_to_owned!(Mul, mul, MatZ);
+arithmetic_trait_mixed_borrowed_owned!(Mul, mul, MatZ);
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/mat_z/mul.rs:60]: We test the borrowed * borrowed function and as the macro just calls this, we just ensure that it works properly by the doc comment of the function.

File: src/integer/mat_z/transpose.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    /// Checks if a column is correctly converted to a row
+    #[test]
+    fn column_to_row() {
+        let mat = MatZ::from_str("[[1,2,3]]").unwrap();
+        let cmp = MatZ::from_str("[[1],[2],[3]]").unwrap();
+
+        assert_eq!(cmp, mat.transpose());
+    }
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/mat_z/transpose.rs:59]: Done

File: src/integer/vec_z/from.rs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        }
+    }
+}
+
+impl FromStr for VecZ {
+    type Err = MathError;
+
+    /// Creates a [`VecZ`] row-vector with entries in [`Z`] from a [`String`].
+    /// The format of that string looks like this `[1,2,3]` for a row vector
+    /// with three entries (`1` in the first row, `2` in the second one, ...)
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/from.rs:66]: does not fit to direction

File: src/integer/vec_z/from.rs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    /// Ensure that entries of a new vector are `0`.
+    #[test]
+    fn entry_zero() {
+        let matrix = VecZ::new(2, VectorDirection::ColumnVector).unwrap();
+
+        let entry1 = matrix.get_entry(0).unwrap();
+        let entry2 = matrix.get_entry(0).unwrap();
+
+        assert_eq!(Z::from_i64(0), entry1);
+        assert_eq!(Z::from_i64(0), entry2);
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/from.rs:132]:
```suggestion
        assert_eq!(Z::ZERO entry1);
        assert_eq!(Z::ZERO, entry2);
```

Comment (Marvin-Beckmann) [inline src/integer/vec_z/from.rs:132]: though i prefer using the constant
```suggestion
        assert_eq!(Z::from(0), entry1);
        assert_eq!(Z::from(0), entry2);
```

File: src/integer/vec_z/from.rs [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    use crate::utils::VectorDirection;
+    use std::str::FromStr;
+
+    /// Ensure that initialization works in both directions
+    #[test]
+    fn directions() {
+        let row = VecZ::from_str("[[1, 2, 3]]").unwrap();
+        let col = VecZ::from_str("[[1],[2],[3]]").unwrap();
+
+        assert_eq!(Z::from_i64(1), row.get_entry(0).unwrap());
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/from.rs:159]:
```suggestion
        assert_eq!(Z::ONE, row.get_entry(0).unwrap());
```

File: src/integer/vec_z/from.rs [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+        assert!(row.is_row_vector());
+        assert_eq!(VectorDirection::RowVector, row.get_direction());
+
+        assert_eq!(Z::from_i64(1), col.get_entry(0).unwrap());
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/from.rs:163]:
```suggestion
        assert_eq!(Z::ONE, col.get_entry(0).unwrap());
```

File: src/integer/vec_z/from.rs [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+        );
+    }
+
+    /// Ensure that entries can have leading and trailing whitespaces.
+    #[test]
+    fn whitespaces_in_entries_works() {
+        let vector_string = String::from("[[  1],[2 ],[  3  ]]");
+
+        assert_eq!(
+            Z::from_i64(1),
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/from.rs:204]:
```suggestion
            Z::ONE,
```

File: src/integer/vec_z/from.rs [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+    /// [`InvalidStringToZInput`](MathError::InvalidStringToZInput)
+    /// if an entry is not formatted correctly.
+    fn from_str(string: &str) -> Result<Self, MathError> {
+        let entries_string = parse_matrix_string(string)?;
+        let (num_rows, num_cols) = find_matrix_dimensions(&entries_string)?;
+        let mut vector: VecZ;
+        match (num_rows, num_cols) {
+            (1, _) => vector = VecZ::new(num_cols, VectorDirection::RowVector)?,
+            (_, 1) => vector = VecZ::new(num_rows, VectorDirection::ColumnVector)?,
+            (_, _) => return Err(MathError::InvalidStringToVectorInput(String::from(
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/from.rs:101]: should be possible to wirte it like this
```suggestion
            _ => return Err(MathError::InvalidStringToVectorInput(String::from(
```
else consider including the amount of rows and columns in the error message

File: src/integer/vec_z/from.rs [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+        assert_eq!(vector.get_entry(1).unwrap(), i64::MIN.into());
+        assert_eq!(vector.get_entry(2).unwrap(), i64::MAX.into());
+        assert_eq!(vector.get_entry(3).unwrap(), (-10).into());
+        assert_eq!(vector.get_entry(4).unwrap(), 10.into());
+    }
+
+    /// Ensure that a wrong format causes an error.
+    #[test]
+    fn wrong_format_error() {
+        let vector_string1 = String::from("[1, 2, 3],");
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/from.rs:243]: test for a valid matrix but not vector is missing

File: src/integer/vec_z/mul.rs [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    /// let a = VecZ::from_str("[[2],[1],[1],[2]]").unwrap();
+    /// let b = VecZ::from_str("[[1,0,0,1]]").unwrap();
+    ///
+    /// let _ = &a * &b;
+    /// let _ = a.clone() * b.clone();
+    /// let _ = &a * b.clone();
+    /// let _ = a.clone() * &b;
+    /// ```
+    fn mul(self, other: Self) -> Self::Output {
+        (&self.matrix).mul(other)
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/mul.rs:43]:
```suggestion
        // TODO: mul_safe
        (&self.matrix).mul(other)
```

File: src/integer/vec_z/set.rs [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    #[test]
+    fn correct_entry_set() {
+        let mut col = VecZ::new(5, VectorDirection::ColumnVector).unwrap();
+        let mut row = VecZ::new(5, VectorDirection::RowVector).unwrap();
+
+        col.set_entry(4, 869).unwrap();
+        row.set_entry(3, 869).unwrap();
+
+        assert_eq!(col.get_entry(4).unwrap(), Z::from_i64(869));
+        assert_eq!(row.get_entry(3).unwrap(), Z::from_i64(869));
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/set.rs:93]:
```suggestion
        assert_eq!(col.get_entry(4).unwrap(), Z::from(869));
        assert_eq!(row.get_entry(3).unwrap(), Z::from(869));
```

File: src/integer/vec_z/set.rs [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    /// Ensure that a wrong given row returns a [`MathError`](crate::error::MathError)
+    #[test]
+    fn error_wrong_row() {
+        let mut col = VecZ::new(5, VectorDirection::ColumnVector).unwrap();
+        let mut row = VecZ::new(5, VectorDirection::RowVector).unwrap();
+        let value = Z::from_i64(i64::MAX);
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/set.rs:101]:
```suggestion
        let value = Z::from(i64::MAX);
```

File: src/integer/vec_z/set.rs [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    /// - `entry`: specifies the entry in which the entry is located
+    /// - `value`: specifies the value to which the entry is set
+    ///
+    /// # Example
+    /// ```
+    /// use math::integer::{VecZ, Z};
+    /// use math::utils::VectorDirection;
+    ///
+    /// let mut vector = VecZ::new(5, VectorDirection::ColumnVector).unwrap();
+    /// let value = Z::from_i64(5);
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/set.rs:55]:
```suggestion
    /// let value = Z::from(5);
```

File: src/integer/vec_z/transpose.rs [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /// Checks if a vector is correctly converted to a one-column matrix
+    #[test]
+    fn row_to_column() {
+        let vec = VecZ::from_str("[[1,2,3]]").unwrap();
+        let cmp = VecZ::from_str("[[1],[2],[3]]").unwrap();
+
+        assert_eq!(cmp, vec.transpose());
+    }
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/transpose.rs:57]: Please also test with large entries and with negative entries and different sizes

File: src/integer/vec_z/from.rs [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+        }
+
+        // fill entries of matrix according to entries in string_matrix
+        for (row_num, row) in entries_string.iter().enumerate() {
+            for (col_num, entry) in row.iter().enumerate() {
+                let z_entry = Z::from_str(entry)?;
+                vector.matrix.set_entry(row_num, col_num, z_entry)?;
+            }
+        }
+        Ok(vector)
```
Comment (Marvin-Beckmann) [inline src/integer/vec_z/from.rs:113]: Alternatively, what I would prefer:
```suggestion
        let matrix = MatZ::from_str(str)?;
        match (matrix.get_num_rows(), matrix.get_num_cols()) {
            (a,b) if a ==1 || b== 1 => Ok(VecZ{matrix})}
            _ => Err(MathError::InvalidStringToVectorInput(String::from(
                "The string either contained more than one column or row, or did have zero entries",
            ))),
        }
```

File: src/integer/vec_z/set.rs [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+    /// vector.set_entry(1, value).unwrap();
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of type [`OutOfBounds`](MathError::OutOfBounds)
+    /// if the given entry outside the vector or negative.
+    pub fn set_entry<S: TryInto<i64> + Display + Copy, T: Into<Z>>(
+        &mut self,
+        entry: S,
+        value: T,
```
Comment (Marcel583) [inline src/integer/vec_z/set.rs:38]: Use updated version with impl instead of S and T

File: src/integer/vec_z/from.rs (hunk 17 above)
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:113]: Done in slightly different way

File: src/integer/vec_z/from.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:66]: Should now fit, again

File: src/integer/vec_z/get.rs [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+    /// use math::integer::VecZ;
+    /// use std::str::FromStr;
+    ///
+    /// let row = VecZ::from_str("[[1,2,3]]").unwrap();
+    ///
+    /// assert!(row.is_row_vector());
+    /// ```
+    pub fn is_row_vector(&self) -> bool {
+        self.matrix.get_num_columns() > 1
+    }
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/get.rs:62]: Now done this way. get_direction changed appropriately

File: src/integer/vec_z/transpose.rs (hunk 16 above)
Comment (jnsiemer) [reply src/integer/vec_z/transpose.rs:57]: several kinds of entries (large, negative etc) are tested in MatZ::transpose, which this function uses. Different sized matrices were now added (4 and 5 entries) to the ones in MatZ:: transpose by these tests

File: src/integer/mat_z/mul.rs [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+    /// use std::str::FromStr;
+    ///
+    /// let a = MatZ::from_str("[[2,1],[1,2]]").unwrap();
+    /// let b = MatZ::from_str("[[1,0],[0,1]]").unwrap();
+    ///
+    /// let c = &a * &b;
+    /// let d = a * b;
+    /// let e = &c * d;
+    /// let f = c * &e;
+    /// ```
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/mat_z/mul.rs:46]: Added

File: src/integer/vec_z/mul.rs [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+    /// use std::str::FromStr;
+    ///
+    /// let a = VecZ::from_str("[[1],[0]]").unwrap();
+    /// let b = MatZ::from_str("[[2,1]]").unwrap();
+    ///
+    /// let c = &a * &b;
+    /// let d = a.clone() * b.clone();
+    /// let e = &a * b.clone();
+    /// let f = a.clone() * &b;
+    /// ```
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/mul.rs:73]: Done

File: src/integer/vec_z/mul.rs [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+    fn mul(self, other: &MatZ) -> Self::Output {
+        (&self.matrix).mul(other)
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/mul.rs:75]: Why would I call mul_safe here if another function already provides exactly the behaviour I want to establish for this function.

File: src/integer/vec_z/mul.rs (hunk 12 above)
Comment (jnsiemer) [reply src/integer/vec_z/mul.rs:43]: Why would I call mul_safe here if another function already provides exactly the behaviour I want to establish for this function.

File: src/integer/vec_z/from.rs (hunk 9 above)
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:204]: Done

File: src/integer/vec_z/from.rs (hunk 7 above)
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:159]: Done

File: src/integer/vec_z/from.rs (hunk 6 above)
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:132]: Changed to Z::ZERO

Comment (jnsiemer) [reply src/integer/vec_z/from.rs:132]: Done

File: src/integer/mat_z/cmp.rs [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+// This is not guaranteed by the [`PartialEq`] trait.
+impl Eq for MatZ {}
+
+/// Test that the [`PartialEq`] trait is correctly implemented.
+#[cfg(test)]
+mod test_partial_eq {
+
+    use super::MatZ;
+    use std::str::FromStr;
+
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/mat_z/cmp.rs:55]: Now one 3x3 test was included

File: src/utils/parse.rs [hunk 24]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2023 Marcel Luca Schmidt
+// Copyright © 2023 Marcel Luca Schmidt, Niklas Siemer
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/utils/parse.rs:1]: Done

File: src/integer/mat_z/transpose.rs [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    /// use math::integer::MatZ;
+    /// use std::str::FromStr;
+    ///
+    /// let mat = MatZ::from_str("[[2,1],[2,1],[2,1]]").unwrap();
+    /// let cmp = MatZ::from_str("[[2,2,2],[1,1,1]]").unwrap();
+    ///
+    /// assert_eq!(mat.transpose(), cmp);
+    /// ```
+    pub fn transpose(&self) -> Self {
+        let mut out = Self::new(self.get_num_columns(), self.get_num_rows()).unwrap();
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/mat_z/transpose.rs:31]: As we know that this can not fail, it's ok to use unwrap

File: src/integer/vec_z/from.rs [hunk 26]
```diff
@@ -0,0 +1,263 @@
...
+    /// use math::integer::VecZ;
+    /// use math::utils::VectorDirection;
+    ///
+    /// let vector = VecZ::new(5, VectorDirection::ColumnVector).unwrap();
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of type
+    /// [`InvalidMatrix`](MathError::InvalidMatrix)
+    /// if the number of rows is `0`.
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:43]: Done

File: src/integer/vec_z/from.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:66]: Done, but left out the comment about a 1x1 string as this is the wrong place to put it.

File: src/integer/vec_z/from.rs [hunk 27]
```diff
@@ -0,0 +1,263 @@
...
+
+    /// Creates a [`VecZ`] row-vector with entries in [`Z`] from a [`String`].
+    /// The format of that string looks like this `[1,2,3]` for a row vector
+    /// with three entries (`1` in the first row, `2` in the second one, ...)
+    ///
+    /// Parameters:
+    /// - `string`: the vector representation as a string
+    ///
+    /// Returns a [`VecZ`] or an error, if the vector is not formatted in a suitable way,
+    /// the number of rows is too big (must fit into [`i64`]), or if the regular expression
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:72]: Done

File: src/integer/vec_z/from.rs [hunk 28]
```diff
@@ -0,0 +1,263 @@
...
+    /// inside of the function could not be processed.
+    ///
+    /// # Example
+    /// ```
+    /// use math::integer::VecZ;
+    /// use std::str::FromStr;
+    ///
+    /// let string = String::from("[[1, 2, 3]]");
+    /// let matrix = VecZ::from_str(&string).unwrap();
+    /// ```
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:82]: Done

File: src/integer/vec_z/from.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:204]: Done

[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:204]: Done

File: src/integer/vec_z/from.rs [hunk 29]
```diff
@@ -0,0 +1,263 @@
...
+
+        assert_eq!(Z::from_i64(1), col.get_entry(0).unwrap());
+        assert!(col.is_column_vector());
+        assert_eq!(VectorDirection::ColumnVector, col.get_direction());
+    }
+
+    /// Ensure that initialization with positive numbers that are larger than [`i64`] works.
+    #[test]
+    fn large_numbers() {
+        let vector_string = format!("[[{}, 2, 3]]", "1".repeat(65));
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/from.rs:171]: Done

File: src/integer/vec_z/get.rs [hunk 30]
```diff
@@ -0,0 +1,193 @@
...
+    /// if the given entry is outside the vector or negative.
+    pub fn get_entry(&self, entry: impl TryInto<i64> + Display + Copy) -> Result<Z, MathError> {
+        if self.is_column_vector() {
+            self.matrix.get_entry(entry, 0)
+        } else {
+            self.matrix.get_entry(0, entry)
+        }
+    }
+
+    /// Returns `true` if the vector is a row vector with more than one column.
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/get.rs:49]: Changed method functionality completely, hence doc comment changed accordingly.

File: src/integer/vec_z/get.rs [hunk 31]
```diff
@@ -0,0 +1,193 @@
...
+    ///
+    /// let row = VecZ::from_str("[[1,2,3]]").unwrap();
+    ///
+    /// assert!(row.is_row_vector());
+    /// ```
+    pub fn is_row_vector(&self) -> bool {
+        self.matrix.get_num_columns() > 1
+    }
+
+    /// Returns `true` if the vector is a column vector with more than one row.
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/get.rs:64]: Changed according to new implementation

File: src/integer/vec_z/get.rs [hunk 32]
```diff
@@ -0,0 +1,193 @@
...
+    /// let col = VecZ::from_str("[[1],[2],[3]]").unwrap();
+    ///
+    /// assert!(col.is_column_vector());
+    /// ```
+    pub fn is_column_vector(&self) -> bool {
+        self.matrix.get_num_rows() > 1
+    }
+
+    /// Returns [`VectorDirection::ColumnVector`] if the vector is a column vector with more than one row.
+    /// Otherwise, returns [`VectorDirection::RowVector`].
```
[the remark this answers is not in the document]
Comment (jnsiemer) [reply src/integer/vec_z/get.rs:80]: Changed up completely

File: src/integer/vec_z/set.rs (hunk 18 above)
Comment (jnsiemer) [reply src/integer/vec_z/set.rs:38]: As I'm not able to use the trait and this is the way it is done in MatZ, I do not see what to do differently here. But I also opened an issue for the set_entry and get_entry trait for Vectors.

**Review by Marvin-Beckmann — APPROVED** [review]

**Review by Marcel583 — APPROVED** [review]
